### ISSUE TITLE
[ttnn.sampling] RNG path lacks per-row entropy, per-step entropy, and runtime seed support

### ISSUE THREAD

**Issue by kmabeeTT** [tenstorrent/tt-metal#44661]
# [ttnn.sampling] RNG path lacks per-row entropy, per-step entropy, and runtime seed support

## Summary

`ttnn.sampling`'s RNG plumbing has three independent functional gaps that together make it impossible to honor `SamplingParams(seed=...)` in a vLLM-style sampler:

| Required behavior | Current `ttnn.sampling` | Bug |
| --- | --- | --- |
| Independent entropy per row in a batched call | Scalar seed broadcast to all 32 cores; every core's SFPU RNG initializes identically | **A** |
| Entropy advances across consecutive calls | `rand_tile_init(seed)` re-runs every invocation when `seed != 0`, resetting RNG state | **B** |
| Changing seed at runtime without a kernel recompile | `seed` is a kernel compile-time arg; each new seed value forces a program-cache miss + recompile | **C** |

These limitations forced tt-xla to (a) reject `SamplingParams(seed=...)` on the device sampler with a clear error (`TTPlatform.validate_request()` in `integrations/vllm_plugin/vllm_tt/platform.py:209-231`, points users at `additional_config={"cpu_sampling": True}`), and (b) `xfail` the synthetic seed tests in `tests/integrations/vllm_plugin/sampling/test_sampling_params_synthetic.py::test_seed*`. Once the kernel grows the missing functionality, both workarounds come off. tt-xla tracking issue: tenstorrent/tt-xla#4539.

Confirmed on tt-metal `78197e564f3` (2026-05-04, the version tt-xla currently consumes).

## Impact / priority

There is no viable workaround at the tt-xla level. In principle we could route seeded requests through tt-xla's previous (non-fused) Gumbel-max device sampler — the code path still exists; it's what `cpu_sampling=True` uses on the host today. We considered this and rejected it: a surprise perf cliff triggered by setting `SamplingParams(seed=...)` is worse UX than a clean validation error, since users would observe inconsistent throughput depending on a sampling knob and have no clear way to debug it. The current posture (`validate_request` raises, pointing users at `cpu_sampling=True`) is the deliberate choice, not a temporary hack. That makes this kernel fix the unblock for honoring `SamplingParams(seed=...)` on the device sampler at all — there is no consumer-side path that gets us there without it.

## Evidence

**Bug A — scalar seed shared across all 32 cores; no per-row entropy.**

`ttnn/cpp/ttnn/operations/reduction/sampling/device/sampling_program_factory.cpp:315-316` registers one `compute_args` vector (containing `random_seed`) against every core in the grid:

```cpp
std::vector<uint32_t> compute_args = { ..., random_seed, ... };
```

`ttnn/cpp/ttnn/operations/reduction/sampling/device/kernels/compute/sampling.cpp:26-48`:

```cpp
void generate_rand_tile(const uint32_t cb_id, const uint32_t seed) {
    init_sfpu(cb_id, cb_id);
    ...
    if (seed != 0) {
        rand_tile_init(seed);
    }
    ...
    rand_tile(0, rand_from, rand_scale);
    ...
}
```

No core index, no row index, no per-core entropy source. With identical inputs across all 32 batch rows, the kernel produces 1 distinct sample where independent draws would produce ~28-32 (birthday-problem expectation).

**Bug B — `rand_tile_init` resets RNG state every invocation when `seed != 0`.**

Same lines 34-36: every call re-runs `rand_tile_init(seed)`, resetting each core's SFPU RNG to the same starting state. Two back-to-back calls with the same seed draw byte-identical random tiles. In an autoregressive decode loop, every token step pulls the same RNG draw against (different) logits — the random offset never advances.

**Bug C — seed is a kernel compile-time arg.**

`ttnn/cpp/ttnn/operations/reduction/sampling/device/kernels/compute/sampling.cpp:368`:

```cpp
constexpr uint32_t seed = get_compile_time_arg_val(15);
```

Plumbed from `ttnn/cpp/ttnn/operations/reduction/sampling/device/sampling_program_factory.cpp:67-68`:

```cpp
if (seed.has_value()) {
    random_seed = seed.value();   // becomes a compile-time arg below
}
```

`random_seed` is part of the program-factory hash → each distinct seed value triggers a program-cache miss and a full kernel recompile. This blocks runtime seed toggling, which is required to honor per-request `SamplingParams.seed` in a vLLM-style sampler (every request can carry its own seed, and the kernel cannot be expected to rebuild on each one).

Measured cost — `bench_decode_loop.py` on Blackhole, tt-metal `78197e564f3`. 32 sequential `ttnn.sampling` calls at the candidate-set widths tt-xla / vLLM passes for real models (`W = num_chunks(vocab) × 32`, padded to power-of-2):

| Model surface           | W (candidates) | Baseline (`seed=0`) | Unique seed per call | Per-call recompile cost |
| ----------------------- | -------------- | ------------------- | -------------------- | ----------------------- |
| TinyLlama-1.1B-Chat     | 64             | 0.15 ms             | 1759.31 ms           | **+1.76 s**             |
| Llama-3.1-8B-Instruct   | 128            | 0.16 ms             | 1771.12 ms           | **+1.77 s**             |

The cost is dominated by kernel build/upload, not kernel size — both widths land at ~1.77 s. Either option in the Suggested Fixes section below resolves this — a runtime seed arg or moving entropy out of the kernel entirely via a noise tensor input.

## Repro

Standalone TTNN — no tt-mlir / tt-xla / vLLM. Four self-checking cases, exits non-zero on any failure.

```bash
python repro_ttnn_sampling_seed.py
```

Cherry-pick [`repro_ttnn_sampling_seed_issue44661.py`](https://github.com/tenstorrent/tt-metal/blob/kmabee/repro_ttnn_sampling_seed_issue_44661/repro_ttnn_sampling_seed_issue44661.py) from branch `kmabee/repro_ttnn_sampling_seed_issue_44661` (commit [`1348964a98`](https://github.com/tenstorrent/tt-metal/commit/1348964a98)). The file sits at the repo root; no build needed.

The script doubles as a verification suite: any of the kernel-side fixes in the Suggested Fixes section below should flip the corresponding case from FAIL to PASS, and a complete fix exits 0 ("All cases PASS"). Failure output identifies which sub-bug is still active by case ID.

Captured output (Blackhole, tt-metal `78197e564f3`; kernel code is hardware-agnostic so Wormhole should behave identically):

```
=== Case A: seed=0, uniform distribution -> all rows correlated ===
  samples (32 rows): [19, 17, 19, 19, 17, 17, 19, 19, 17, 17, 17, 17, 17, 19, 17, 19, ...]
  distinct values: 2
  expected (proper per-row RNG): ~28-32
  -> FAIL (bug A reproduced)

=== Case B: seed=42 -> all 32 rows byte-identical ===
  samples (32 rows): [19, 19, 19, 19, 19, 19, 19, 19, 19, 19, 19, 19, 19, 19, 19, 19, ...]
  distinct values: 1
  -> FAIL (bug A reproduced; rand_tile_init resets all cores identically)

=== Case C: seed=42 twice -> second call byte-identical to first ===
  call 1 samples: [19, 19, 19, 19, ..., 19]
  call 2 samples: [19, 19, 19, 19, ..., 19]
  -> FAIL (bug B reproduced; rand_tile_init called every invocation)

=== Case D: seed is compile-time -> each new seed triggers recompile ===
  seed=42 (cold):            0.2 ms
  seed=42 (warm cache):      0.2 ms
  seed=43 (new seed):       14.8 ms     (~74x slower than warm cache)
  seed=44 (new seed):       14.8 ms
  -> FAIL (bug C reproduced; new seed = recompile)
```

On the first run after a cold compile cache, Case D's new-seed calls measured ~1780 ms each (~9000× slower than warm cache).

## Suggested Fixes

Two viable paths. We don't have a strong preference between them from the tt-xla side — both are roughly equivalent on performance and on the amount of consumer-side code change needed. The choice is mostly a design preference for tt-metal: which approach better fits the op's intended architecture and the conventions of similar ttnn ops. Worth a discussion before either is picked.

### Option A: accept a host-precomputed `noise` tensor input

Add an optional `noise` tensor to `ttnn.sampling`, shape `[32]`, bf16, ROW_MAJOR (same convention as `k`/`p`/`temp`). When provided, the writer kernel reads its core's value from this tensor instead of from the internally-generated random tile. The kernel-side `generate_rand_tile` path can stay as the default for callers that don't supply noise.

Matches the convention used by JAX (`jax.random.categorical(key, ...)`), PyTorch (`torch.multinomial(probs, generator=...)`), and tt-xla's CPU sampling path (`Sampler.random_sample(probs, generators, q_samples)`). vLLM already computes a per-row, per-step `q_samples` tensor for CPU sampling; threading that into the device path is mechanical once the kernel accepts it. Sidesteps all three sub-bugs in one shot — entropy lives in the data, host-side, where it's naturally per-row and per-step.

A proof-of-concept implementation is on branch [`kmabee/proposed_fixes_4539_noise_tensor`](https://github.com/tenstorrent/tt-metal/tree/kmabee/proposed_fixes_4539_noise_tensor) (commit `160e40c8d94`, 8 files, +79/-13). Compiles, validates 3 of 4 repro cases. The 4th case fails on a framework subtlety the ttnn team will likely resolve quickly — see the commit message for details.

### Option B: fix the internal RNG path

Keep the kernel-internal RNG but address all three sub-bugs:

- **Bug A**: derive a per-core seed (e.g. `effective_seed = seed XOR splitmix(core_idx)`) so each row has independent entropy.
- **Bug B**: either don't re-init the SFPU RNG on subsequent calls, or accept a `step_index` runtime arg so the kernel can derive a per-call sub-seed.
- **Bug C**: move `seed` from a kernel compile-time arg to a runtime arg, and exclude `seed` from `compute_program_hash` so a new seed value reuses the existing program-cache entry.

Matches the convention used by `ttnn.bernoulli`, `ttnn.rand`, `ttnn.uniform`, and the other reduction/random ops in tt-metal — caller passes a scalar seed, kernel does the RNG. No host-device data transfer for noise.

A proof-of-concept implementation of the Bug A + Bug C parts is on branch [`kmabee/proposed_fixes_4539_internal_rng`](https://github.com/tenstorrent/tt-metal/tree/kmabee/proposed_fixes_4539_internal_rng) (commit `0bb9523a018`, 4 files, +40/-10). Compiles cleanly, validates A, B, D cases of the repro from FAIL to PASS. Bug B is the harder one to address on this path; it needs a `step_index` runtime arg or equivalent per-launch state.

### Trade-off summary

| Concern | Option A (noise tensor) | Option B (internal RNG fix) |
| --- | --- | --- |
| Per-call overhead vs current | +host RNG (~3-10 μs) + DMA (~5-20 μs) | +per-core seed derivation (~10 cycles) |
| Multi-chip TP coordination | Host broadcasts noise — trivially consistent | Same seed across chips produces same RNG draws — naturally consistent |
| Matches existing convention of | `ttnn.bernoulli` / `ttnn.rand` / etc. on tt-metal: No. JAX / PyTorch / vLLM CPU path: Yes. | `ttnn.bernoulli` / `ttnn.rand` / etc.: Yes. JAX-style key plumbing: No. |
| Closes all three sub-bugs by design | Yes (entropy lives in data) | Yes, but Bug B needs the harder framework piece (`step_index` runtime arg) |
| Op-signature change | New optional `noise` tensor input | None on the API; just internal arg changes (mostly) |
| Implementation status | WIP branch, framework subtlety blocks final case | Verified working for 3 of 4 sub-bugs |

For context: tt-xla's prior (non-fused) sampler used Option A semantics (per-row, per-step `q_samples` tensor). Adopting `ttnn.sampling` lost that flexibility, which is why we're filing this.

Once either path lands, tt-xla will thread the appropriate plumbing (noise tensor or per-request seed) into the device sampler, drop the `validate_request` rejection, and un-xfail the synthetic seed tests.

## Environment

- tt-metal: `78197e564f3` (2026-05-04, the version tt-xla currently consumes)
- Device: Reproduced on single-chip Blackhole; kernel code is hardware-agnostic so Wormhole should behave identically (tt-xla's vLLM stack hits the same kernel on Wormhole n300 and n300_llmbox)

**Comment by sraizada-tt** [comment]
tt-metal vLLM dodges all three bugs by using a separate composition around `ttnn.sampling` rather than the `seed=` kwarg, so our CI runs are not affected by this. Sharing the pattern in case it's useful context for the Option A vs Option B design discussion — it's roughly Option B's runtime-seed semantics assembled at the host layer today.

Path lives in `models/common/sampling/tt_sampling.py::TTSampling` + `models/common/sampling/generator.py::SeedManager`. Every step is a two-call sequence:

```python
# One uint32 RNG slot per user, runtime device tensor:
self.seeds_tt_tensor = ttnn.from_torch(
    torch.arange(total_param_size).to(torch.uint32),
    dtype=ttnn.uint32,
    ...
)

# Each step, before ttnn.sampling:
ttnn.manual_seed(
    seeds=self.seeds_tt_tensor,
    user_ids=self.user_ids_tt_tensor,
    sub_core_grids=self.sub_core_grids,
)
tt_out_tok = ttnn.sampling(..., output_tensor=tt_out_tok)   # no seed= kwarg
```

`ttnn.manual_seed` (`ttnn/cpp/ttnn/operations/reduction/manual_seed/`) is a small op whose reader matches `core_id` against `user_ids` and writes that user's seed into per-core L1, where the next `ttnn.sampling` call picks it up.

Against your three sub-bugs:

- **A (per-row entropy)**: each core looks up its own user's seed slot from the device tensor, so the 32 batch rows initialise the SFPU RNG to distinct states.
- **B (state advances per step)**: `SeedManager` keeps a `random.Random` per user host-side. For seeded slots it advances the host RNG and pushes the new value into `seeds_tt_tensor` every step → `rand_tile_init` reseeds with a fresh value each call. For unseeded slots, a three-state machine pushes `MAX_UINT32` after the initial diversifying push; the compute kernel's `if (seed != 0)` guard treats that as a SKIP sentinel and lets `rand_tile` advance the existing per-core state instead of re-initing.
- **C (no recompile per seed)**: with `ttnn.sampling` called without a `seed` kwarg, the kernel's compile-time arg stays constant (0). All per-request entropy travels through the runtime `seeds_tt_tensor` and `manual_seed` op. No program-cache miss on seed change.

**Comment by feslamiTT** [comment]
Should this issue be addressed to Ops, @sraizada-tt, @djordje-tt, @cmaryanTT?

**Comment by feslamiTT** [comment]
@kmabeeTT do you still need this fix?

**Comment by mtairum** [comment]
Is this still needed?
